thecourseforum/theCourseForum2
# Lous list open seats

## GitHub Issues addressed

- This PR closes

## Testing

- A brief explanation of tests done/written or how reviewers can test your work

**Note: Ensure that you have the most recent database backup. Additionally, make sure to run `docker build --no-cache` after checking out this branch.**

Reviewers can test my work by modifying the `CELERY_BEAT_SCHEDULE` in `tcf_core/settings/base.py` to a time a few minutes ahead of their current time. After making the change, follow these steps:

1. Run the following command:
```bash
docker-compose restart celery_worker celery_beat
```
2. View the logs by running:
```bash
docker-compose logs -f celery_worker
```
3. This process will populate the database with the correct enrollment records. Once populated (or at least some of the sections are populated), you can test the filters to query based on open sections. Be sure to test applying the filters in both the dropdown search menu and the advanced search menu to confirm that the functionality works as intended.

## What I did

- Created SectionEnrollment model to track course section enrollment
- Implemented celery task to update enrollment and waitlist data daily at 4 AM EST
- Added open sections filter to course search functionality
- Integrated open sections filter into dropdown and advanced search UI
- Updated Course model and search view to handle open sections filtering

## Screenshots

- Before
<img width="867" alt="image" src="https://github.com/user-attachments/assets/b00be5e6-0eed-42b2-92a1-e8fbdfeb8515">
<img width="562" alt="image" src="https://github.com/user-attachments/assets/4944a53d-0209-49d6-bac7-c55c8319f07d">
<img width="587" alt="image" src="https://github.com/user-attachments/assets/5ffe4a13-c414-4e64-a5e0-d88fa50c24fc">

*Open Section Filter Not Applied*
<img width="553" alt="image" src="https://github.com/user-attachments/assets/f5ff9fa3-c671-4cdc-b19c-f4641febcbca">
- After
<img width="856" alt="image" src="https://github.com/user-attachments/assets/002a0de3-ec59-4ffe-9f91-1c00776100de">
<img width="514" alt="image" src="https://github.com/user-attachments/assets/b40b2bd6-14e5-4b4b-b4fd-08f1471747b8">
<img width="584" alt="image" src="https://github.com/user-attachments/assets/330c3871-83ab-4638-ad3c-75aad6084808">

*Open Section Filter Applied*
<img width="557" alt="image" src="https://github.com/user-attachments/assets/da749753-a027-403e-9bee-12a68b5e025c">


## Questions/Discussions/Notes

- I'm encountering a bug on the advanced search page. When I select the checkbox directly, it behaves as expected by applying the filter and does not appear as a selected filter in the display on the top of the screen. However, when I click on the label associated with the checkbox, instead of selecting the checkbox, it applies the filter to the display at the top of the screen. Despite this behavior, the filters are being correctly applied in both cases and the search functions as expected.

Nothing Selected:
<img width="1124" alt="image" src="https://github.com/user-attachments/assets/5763ba6c-1d37-4588-81db-0a6914c5bcd1">

Selected Checkbox:
<img width="1124" alt="image" src="https://github.com/user-attachments/assets/53e4e649-d311-44be-9578-724a0ac3798e">

Selected Label:
<img width="1127" alt="image" src="https://github.com/user-attachments/assets/055984b1-8c79-4e45-81ad-6dcd503a82d4">

**Intended Functionality:**

**Advanced Search Page:**
- Clicking either the checkbox or its label should select the checkbox.
- The filter should only appear in the top search bar after submitting a query.

**Dropdown Search Menu:**
- This functions correctly, showing the filter in the top search bar when selected. It does not need to be modified.
